docs/DesignDocs/AttestationQPLEnforcementSupport.md

```diff
@@ -0,0 +1,95 @@
+Attestation SGX QPL Enforcement Support
+====
+
+This proposal is to make a change to `oe_verify_evidence`, so that it can support policy-based SGX QPL (Quote Provider Library) specific enhancement, such as allowing the API user to pass data to the SGX QPL for more advanced quote verification.
+
+Motivation
+----
+
+Intel SGX rolls out PCK certificate and collateral information about every six months.
+There is a scenario where a quote signed by the last version of PCK certificate cannot be verified with the latest version of collateral information.
+This causes unnecessary attestation failures when new versions of PCK certificate and collateral information are rolled out.
+To deal with this issue, the SGX QPL need to return any version of collateral information based on additional query parameters (referred to as "baselines").
+With that, a new API called `sgx_ql_get_quote_verification_collateral_with_params` was introduced in the SGX QPL.
+Changes are also needed in the OE SDK to call this new API.
+
+Goals:
+ - Verifier can call `oe_verify_evidence` and pass SGX QPL-specific parameters to do more advanced quote verification.
+
+User Experience
+----
+
+A new policy type `OE_POLICY_SGX_COLLATERAL_BASELINE` is added to current policy types.
+A quote verifier application can pass multiple policies with this type, while OE SDK would pass all the policies down to the SGX QPL API without interpreting the data content.
+
+Specification
+----
+
+### New policy type & its data structure
+
+- `OE_POLICY_SGX_COLLATERAL_BASELINE` in `oe_policy_type_t`
```

Comment:
Now that the type has a TEE-agnostic name (oe_query_param_t), could the policy type value also have a more TEE-agnostic name?
E.g., OE_POLICY_OPTIONS or something?

Reply:
I thought about that before, see the "Alternatives", the problem is OE SDK needs to know how to route the parameters, for generalized the oe_query_param_t, I would take it as OE_POLICY_SGX_COLLERAL_BASELINE data is oe_query_param_t, and in future, some other policies can share the same data type, because it's generic, it does not necessary to have a generic policy name for the generic data type, how do you think?

Reply:
I totally get that there may be multiple policy types with the same struct, but can you say more about what you mean by "how to route the parameters"?  The format UUID in the oe_verify_evidence() call already tells OE that it's evidence to be routed to the SGX plugin to interpret.

Reply:
it's true that UUID points to a provider, however within the oe_verify_evidence call, we may need to call multiple APIs from provider to fulfill the request, now, the question is should we pass all the provider polices to all provider APIs without filtering? or in other words, will all provider APIs take same format of additional parameters?

Reply:
Oh, sorry, just realize that are you saying that the name of the policy does not need to have SGX in it?

Reply:
Are you saying that the SGX plugin needs to call multiple "providers"?  (I'm not sure what a "provider" is).  I'm just saying that in your sample code, lines 69-71 identify to OE that it should be routed to the SGX plugin.   Lines 60 and 63 could have called it OE_POLICY_OPTIONS or whatever else that isn't SGX specific, and I think it would still be routed to the SGX plugin because of lines 69-71, right?  Let me know if I'm confused :)

Reply:
not multiple providers, but multiple APIs in same provider, that is possible, right? if we make the policy to be generic, then we need to enforce all APIs in the same provider to take same set of additional parameters.

Reply:
Reopen this as Dave may heave other questions. @dthaler , I would like to clarify what I referred multiple APIs in the same provider, today, oe_verify_evidence needs to call provider to complete the API call, I think we all agree on this. I assume we also agree on a provider may provide many APIs to OE SDK. Based on that, my point is for a given OE SDK API, it does not necessary to have an enforcement that this OE SDK API only can call one provider API behind the scenes. for example, oe_some_api may have the following logic,
- do validation
- do step 1
- call provider API1 to complete step 1
- do step 2
- call provider API2 to complete step 2
- return the result

In this case, if we pass a generic block of parameters for provider, then we need to pass that block of parameters to both API1 and API2 for the provider. Versus, have a more specific policy saying that these parameters are for API1 of the provider, then we don't need to pass the parameters to API2 at step2. 
let me know if that helps to resolve your confusing.

Reply:
According to 2/09/2022's meeting, since we have a consensus on this topic, I will close this discussion